dotCMS/src/main/java/com/dotcms/jobs/business/job/AbstractJob.java

```diff
@@ -0,0 +1,148 @@
+package com.dotcms.jobs.business.job;
+
+import com.dotcms.jobs.business.error.ErrorDetail;
+import com.dotcms.jobs.business.error.RetryStrategy;
+import com.dotcms.jobs.business.processor.ProgressTracker;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.time.LocalDateTime;
+import java.util.Map;
+import java.util.Optional;
+import org.immutables.value.Value;
+
+/**
+ * Abstract interface for an immutable Job class. This interface defines the structure for job
+ * information in the job processing system. The concrete implementation will be generated as an
+ * immutable class named Job.
+ */
+@Value.Style(typeImmutable = "*", typeAbstract = "Abstract*")
+@Value.Immutable
+@JsonSerialize(as = Job.class)
+@JsonDeserialize(as = Job.class)
+public interface AbstractJob {
```

Comment:
I am a little cautious of adding functionality and transformations within the immutable representation of the object preferring a separation of the data from the transformation.   How this has been done by recreating a new immutable does not seem like a bad pattern as such as long as we are aware of the consequences and do not just make this class to too much.    We have a comparison here with Object serialization using the Active Record Pattern, vs the Repository Pattern in ORM.    There is a cost calling nextRetryDelay for example as it has to create a whole new object and if requesting multiple of calls it will create more.    I want to consider whether having a basic interface that just has nextRetryDelay, maxRetries ... as fields and a service class that creates an returns a Job instance with all these populated makes more sense.   If these fields were infrequently accessed and there is a large cost to return them the current pattern makes sense as otherwise, the service would be doing extra work to provide fields that are never accessed.  If this and the other fields data can be easily accessed by the service and there is no extra cost to adding these fields then it is better to not have these provided on demand as they are now.

Reply:
Done, removed extra logic from the `AbstractJob`.